src/connector/src/parser/avro/parser.rs

```diff
@@ -184,53 +184,107 @@ impl AvroParser {
+                    .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
+                match reader.next() {
+                    Some(Ok(v)) => Some(v),
+                    Some(Err(e)) => return Err(RwError::from(ProtocolError(e.to_string()))),
+                    None => {
+                        return Err(RwError::from(ProtocolError(
+                            "avro parse unexpected eof".to_string(),
+                        )));
+                    }
+                }
+            }
         } else {
-            let mut reader = Reader::with_schema(&self.schema, &payload as &[u8])
+            None
+        };
+
+        let avro_key = if let Some(payload) = raw_key {
+            if let Some(resolver) = &self.schema_resolver {
+                let (schema_id, mut raw_payload) = extract_schema_id(payload.as_ref())?;
+                let writer_schema = resolver.get(schema_id).await?;
+                let reader_schema = self.key_schema.as_ref();
+                let value = from_avro_datum(
+                    writer_schema.as_ref(),
+                    &mut raw_payload,
+                    reader_schema.map(|x| &**x),
+                )
                 .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
-            match reader.next() {
-                Some(Ok(v)) => v,
-                Some(Err(e)) => return Err(RwError::from(ProtocolError(e.to_string()))),
-                None => {
-                    return Err(RwError::from(ProtocolError(
-                        "avro parse unexpected eof".to_string(),
-                    )));
+
+                Some(value)
+            } else if let Some(key_schema) = self.key_schema.as_ref() {
+                let mut reader = Reader::with_schema(key_schema, &payload as &[u8])
+                    .map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
+                match reader.next() {
+                    Some(Ok(v)) => Some(v),
+                    Some(Err(e)) => return Err(RwError::from(ProtocolError(e.to_string()))),
+                    None => {
+                        return Err(RwError::from(ProtocolError(
+                            "avro parse unexpected eof".to_string(),
+                        )));
+                    }
                 }
+            } else {
+                None
             }
+        } else {
+            None
         };
 
         // the avro can be a key or a value
-        if let Value::Record(fields) = avro_value {
+        if let Some(Value::Record(fields)) = avro_value {
             let fill = |column: &SourceColumnDesc| {
                 let tuple = match fields.iter().find(|val| column.name.eq(&val.0)) {
-                    None => return Ok(None),
+                    None => {
+                        if self.upsert_primary_key_column_name.as_ref() == Some(&column.name)
+                            && avro_key.is_some()
+                            && self.key_schema.is_some()
```

Comment:
I think it is invalid to have `self.upsert_primary_key_column_name.as_ref() == Some(&column.name)` but `!avro_key.is_some() || !self.key_schema.is_some()`. Shall we make it an error instead of returning `Ok(None)`?

Reply:
This error may be caused by users, which leave the message key empty. The only thing we can do here is to skip this kind of messages. I have make an error log to indicate such case.